roboteam_ai/src/world/World.cpp

```diff
@@ -0,0 +1,244 @@
+        worldMsg.us.push_back(robotMsg);
+    }
+    for (auto &robot : worldData->them) {
+        auto robotMsg = makeWorldRobotMsg(robot);
+        worldMsg.them.push_back(robotMsg);
+    }
+    worldMsg.ball = makeWorldBallMsg(worldData->ball);
+    worldMsg.time = worldData->time;
+    return worldMsg;
+}
+
+bool World::weHaveRobots() {
+    std::lock_guard<std::mutex> lock(worldMutex);
+
+    return worldData == nullptr ? false : !worldData->us.empty();
+}
+
+void World::setWorld(const roboteam_msgs::World &worldMsg) {
+    std::lock_guard<std::mutex> lockMsg(worldMsgMutex);
+    std::lock_guard<std::mutex> lockyLock(worldMutex);
+
+    World::worldMsg = worldMsg;
+    World::worldData = std::make_shared<WorldData>(WorldData(worldMsg));
+
+}
+
+void World::setBall(const Ball &ball) {
+    std::lock_guard<std::mutex> lock(worldMutex);
+
+    World::worldData->ball = ball;
+}
+
+void World::setWorldData(WorldDataPtr &worldData) {
+    std::lock_guard<std::mutex> lockMsg(worldMsgMutex);
+    std::lock_guard<std::mutex> lockyLock(worldMutex);
+
+    World::worldMsg = makeWorldMsg(worldData);
+    World::worldData.reset();
+    World::worldData = worldData;
+}
+
+const roboteam_msgs::World &World::getWorldMsg() {
+    std::lock_guard<std::mutex> lock(worldMsgMutex);
+
+    return worldMsg;
+}
+
+const roboteam_msgs::WorldBall &World::getBallMsg() {
+    std::lock_guard<std::mutex> lock(worldMsgMutex);
+
+    return worldMsg.ball;
+}
+
+const WorldData World::getWorld() {
+    WorldData worldCopy;
+    {
+        std::lock_guard<std::mutex> lock(worldMutex);
+        worldCopy = *worldData.get();
+    }
+    return worldCopy;
+}
+
+std::shared_ptr<Ball> World::getBall() {
+    Ball ballCopy;
+    {
+        std::lock_guard<std::mutex> lock(worldMutex);
+        ballCopy = worldData.get()->ball;
+    }
+    if (ballCopy.exists != 0)
+        return std::make_shared<Ball>(ballCopy);
+
+    return BallPtr(nullptr);
+
+}
+
+std::shared_ptr<Robot> World::getRobotForId(int id, bool ourTeam) {
+    WorldDataPtr worldCopy;
+    {
+        std::lock_guard<std::mutex> lock(worldMutex);
+        worldCopy = worldData;
+    }
+
+    const std::vector<Robot> &robots = ourTeam ? worldCopy->us : worldCopy->them;
+    for (const auto &bot : robots) {
+        if (bot.id == id) {
+            return std::make_shared<Robot>(bot);
+        }
+    }
+    return RobotPtr(nullptr);
+}
+
+std::vector<std::shared_ptr<Robot>> World::getAllRobots() {
+    WorldDataPtr worldCopy;
+    {
+        std::lock_guard<std::mutex> lock(worldMutex);
+        worldCopy = worldData;
+    }
+
+    std::vector<RobotPtr> allRobots;
+    for (auto &robot : worldCopy->us)
+        allRobots.push_back(std::make_shared<Robot>(robot));
+    for (auto &robot : worldCopy->them)
+        allRobots.push_back(std::make_shared<Robot>(robot));
+    return allRobots;
+}
+
+World::RobotPtr World::getRobotClosestToPoint(const Vector2 &point, WhichRobots whichRobots) {
```

Comment:
I would personally merge this with the other function

Reply:
Lukas would disagree

Reply:
yeah changed it, looks better now imo